File: python/invariance/session.py
from __future__ import annotations

import asyncio
from typing import Any, Awaitable, Callable, TypeVar

from ulid import ULID

from .errors import InvarianceError
from .receipt import create_receipt
from .types import Action, Receipt, SessionInfo

T = TypeVar("T")

EnqueueFn = Callable[[Receipt], None]
SessionCreateFn = Callable[[dict[str, Any]], Awaitable[None]]
SessionCloseFn = Callable[[str, str, str], Awaitable[None]]


class Session:
    def __init__(
        self,
        *,
        agent: str,
        name: str,
        id: str | None = None,
        private_key: str | None = None,
        enqueue: EnqueueFn,
        on_create: SessionCreateFn,
        on_close: SessionCloseFn,
    ) -> None:
        self.id = id or str(ULID())
        self.agent = agent
        self.name = name
        self._private_key = private_key
        self._enqueue = enqueue
        self._on_close = on_close
        self._previous_hash = "0"
        self._receipts: list[Receipt] = []
        self._closed = False

        loop = asyncio.get_running_loop()
        self._ready: asyncio.Future[None] = loop.create_future()

        async def _init() -> None:
            try:
                await on_create({"id": self.id, "name": self.name, "agent_id": agent})
                if not self._ready.done():
                    self._ready.set_result(None)
            except Exception as exc:
                if not self._ready.done():
                    self._ready.set_exception(exc)

        loop.create_task(_init())

    @property
    def ready(self) -> asyncio.Future[None]:
        return self._ready

    async def record(self, action: Action) -> Receipt:
        await self._ready
        if self._closed:
            raise InvarianceError("SESSION_CLOSED", f"Session {self.id} is closed")

        receipt = create_receipt(
            session_id=self.id,
            agent=action.get("agent", self.agent),
            action=action["action"],
            input=action.get("input", {}),
            output=action.get("output"),
            error=action.get("error"),
            previous_hash=self._previous_hash,
            private_key=self._private_key,
        )

        self._previous_hash = receipt["hash"]
        self._receipts.append(receipt)
        self._enqueue(receipt)
        return receipt

    async def wrap(
        self,
        action: Action,
        fn: Callable[[], T | Awaitable[T]],
    ) -> dict[str, Any]:
        """Execute fn, then record a receipt with the result or error.

        Returns:
            {'result': T, 'receipt': Receipt}
        """
        result: Any = None
        error: str | None = None
        output: dict[str, Any] | None = None

        try:
            call_result = fn()
            if asyncio.iscoroutine(call_result) or asyncio.isfuture(call_result):
                result = await call_result
            else:
                result = call_result
            if result is not None and isinstance(result, dict):
                output = result
        except Exception as exc:
            error = str(exc)
            receipt = await self.record({**action, "error": error})
            exc.receipt = receipt  # type: ignore[attr-defined]
            raise

        receipt = await self.record({**action, "output": output, "error": error})
        return {"result": result, "receipt": receipt}

    async def end(self, status: str = "closed") -> SessionInfo:
        if self._closed:
            raise InvarianceError("SESSION_CLOSED", f"Session {self.id} is already closed")
        self._closed = True
        close_hash = self._previous_hash
        await self._on_close(self.id, status, close_hash)
        return self.info()

    def get_receipts(self) -> list[Receipt]:
        return list(self._receipts)

    def info(self) -> SessionInfo:
        return {
            "id": self.id,
            "name": self.name,
            "agent": self.agent,
            "status": "closed" if self._closed else "open",
            "receiptCount": len(self._receipts),
            "rootHash": self._receipts[0]["hash"] if self._receipts else None,
            "closeHash": self._previous_hash if self._closed else None,
        }

File: packages/typescript/src/session.ts
import { ulid } from 'ulid';
import { createReceipt } from './receipt.js';
import { InvarianceError } from './errors.js';
import type { Receipt } from './types/receipt.js';
import type { SessionInfo } from './types/session.js';
import type { Action } from './types/config.js';

export type EnqueueFn = (receipt: Receipt) => void;
export type SessionCreateFn = (opts: { id: string; name: string; agent_id?: string }) => Promise<void>;
export type SessionCloseFn = (id: string, status: 'closed' | 'tampered', closeHash: string) => Promise<void>;

export interface SessionOpts {
  agent: string;
  name: string;
  id?: string;
  privateKey?: string;
  enqueue: EnqueueFn;
  onCreate: SessionCreateFn;
  onClose: SessionCloseFn;
}

export class Session {
  readonly id: string;
  readonly agent: string;
  readonly name: string;
  readonly ready: Promise<void>;

  private previousHash = '0';
  private receipts: Receipt[] = [];
  private closed = false;
  private privateKey?: string;
  private enqueue: EnqueueFn;
  private onClose: SessionCloseFn;
  private readyResolve!: () => void;
  private readyReject!: (err: Error) => void;

  constructor(opts: SessionOpts) {
    this.id = opts.id ?? ulid();
    this.agent = opts.agent;
    this.name = opts.name;
    this.privateKey = opts.privateKey;
    this.enqueue = opts.enqueue;
    this.onClose = opts.onClose;

    this.ready = new Promise<void>((resolve, reject) => {
      this.readyResolve = resolve;
      this.readyReject = reject;
    });

    // Fire session creation in background
    opts.onCreate({ id: this.id, name: this.name, agent_id: opts.agent })
      .then(() => this.readyResolve())
      .catch((err) => this.readyReject(err));
  }

  async record(action: Action): Promise<Receipt> {
    await this.ready;
    if (this.closed) throw new InvarianceError('SESSION_CLOSED', `Session ${this.id} is closed`);

    const receipt = await createReceipt({
      sessionId: this.id,
      agent: action.agent ?? this.agent,
      action: action.action,
      input: action.input,
      output: action.output,
      error: action.error,
      previousHash: this.previousHash,
      privateKey: this.privateKey,
    });

    this.previousHash = receipt.hash;
    this.receipts.push(receipt);
    this.enqueue(receipt);
    return receipt;
  }

  async wrap<T>(
    action: Omit<Action, 'output' | 'error'>,
    fn: () => T | Promise<T>,
  ): Promise<{ result: T; receipt: Receipt }> {
    let result: T;
    let error: string | undefined;
    let output: Record<string, unknown> | undefined;

    try {
      result = await fn();
      if (result !== null && result !== undefined && typeof result === 'object') {
        output = result as Record<string, unknown>;
      }
    } catch (err) {
      error = err instanceof Error ? err.message : String(err);
      const receipt = await this.record({ ...action, error });
      throw Object.assign(err instanceof Error ? err : new Error(error), { receipt });
    }

    const receipt = await this.record({ ...action, output, error });
    return { result, receipt };
  }

  async end(status: 'closed' | 'tampered' = 'closed'): Promise<SessionInfo> {
    if (this.closed) throw new InvarianceError('SESSION_CLOSED', `Session ${this.id} is already closed`);
    this.closed = true;
    const closeHash = this.previousHash;
    await this.onClose(this.id, status, closeHash);
    return this.info();
  }

  getReceipts(): readonly Receipt[] {
    return this.receipts;
  }

  info(): SessionInfo {
    return {
      id: this.id,
      name: this.name,
      agent: this.agent,
      status: this.closed ? 'closed' : 'open',
      receiptCount: this.receipts.length,
      rootHash: this.receipts.length > 0 ? this.receipts[0]!.hash : null,
      closeHash: this.closed ? this.previousHash : null,
    };
  }
}
